src/v/cluster/client_quota_serde.h

```diff
@@ -164,32 +166,117 @@ struct entity_value
     std::optional<uint64_t> controller_mutation_rate;
 };
 
+/// entity_value_diff describes the quotas diff for an entity_key
+struct entity_value_diff
+  : serde::
+      envelope<entity_value_diff, serde::version<0>, serde::compat_version<0>> {
+    enum key : int8_t {
+        producer_byte_rate = 0,
+        consumer_byte_rate,
+        controller_mutation_rate,
+    };
+
+    enum operation : int8_t {
+        upsert = 0,
+        remove,
+    };
+
+    struct entry
+      : serde::envelope<entry, serde::version<0>, serde::compat_version<0>> {
+        constexpr entry() noexcept = default;
+        constexpr entry(operation op, key type, uint64_t value) noexcept
+          : op(op)
+          , type(type)
+          , value(value) {}
+        constexpr entry(key type, uint64_t value) noexcept
+          : entry(operation::upsert, type, value) {}
+
+        friend bool operator==(const entry&, const entry&) = default;
```

Comment:
I wonder if value should be ignored if op == remove, like it is for the hash?

Reply:
Yeah that makes sense. I've implemented `operator==` now